pkg/controller/node/node_controller.go

```diff
@@ -794,10 +794,14 @@ func (ctrl *Controller) syncMachineConfigPool(key string) error {
 		// All the nodes that need to be upgraded should have `NodeUpdateInProgressTaint` so that they're less likely
 		// to be chosen during the scheduling cycle.
 		targetConfig := pool.Spec.Configuration.Name
-		if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] != targetConfig {
+		if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] != targetConfig && !checkIfNodeHasInProgressTaint(node) {
 			if err := ctrl.setUpdateInProgressTaint(ctx, node.Name); err != nil {
 				return goerrs.Wrapf(err, "failed applying %s taint for node %s", constants.NodeUpdateInProgressTaint.Key, node.Name)
 			}
+		} else if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] == targetConfig && checkIfNodeHasInProgressTaint(node) {
+			if err := ctrl.removeUpdateInProgressTaint(ctx, node.Name); err != nil {
+				return goerrs.Wrapf(err, "failed removing %s taint for node %s", constants.NodeUpdateInProgressTaint.Key, node.Name)
+			}
```

Comment:
Not sure changing this is actually clearer, but when I read this I had to think it out more like:
```suggestion
		hasInProgressTaint := checkIfNodeHasInProgressTaint(node)
		if node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey] == pool.Spec.Configuration.Name {
			if hasInProgressTaint {
				if err := ctrl.removeUpdateInProgressTaint(ctx, node.Name); err != nil {
					return goerrs.Wrapf(err, "failed removing %s taint for node %s", constants.NodeUpdateInProgressTaint.Key, node.Name)
				}
			}
		} else {
			if !hasInProgressTaint {
				if err := ctrl.setUpdateInProgressTaint(ctx, node.Name); err != nil {
					return goerrs.Wrapf(err, "failed applying %s taint for node %s", constants.NodeUpdateInProgressTaint.Key, node.Name)
				}
			}
		}
```

Reply:
Done.